# 1. Memory Types

## Cache Access Memory

In [1]:
#include <chrono>
#include <iostream>

In [2]:
using namespace std::chrono;

In the following example, we're going to see the difference between interchanging the indices `i` and `j` when accessing the variable `x`. Keep in mind that C++ is **row major**:

#### x[i][j] = i + j   <- for row major

In [3]:
const int size = 4000;
static int x[size][size];

auto t1 = high_resolution_clock::now(); // start time measurement

for (int i = 0; i < size; i++) {
    for (int j = 0; j < size; j++) {
        x[i][j] = i + j;
        //std::cout << &x[i][j] << ": i=" << i << ", j=" << j << std::endl;
    }
}

auto t2 = high_resolution_clock::now(); // stop time measurement

In [4]:
auto duration = duration_cast<std::chrono::microseconds>(t2 - t1).count();
std::cout << "Execution time: " << duration << " microseconds" << std::endl;

Execution time: 23947 microseconds


#### x[j][i] = i + j   <- for column major

In [5]:
const int size = 4000;
static int x[size][size];

auto t1 = high_resolution_clock::now(); // start time measurement

for (int i = 0; i < size; i++) {
    for (int j = 0; j < size; j++) {
        x[j][i] = i + j;
        //std::cout << &x[j][i] << ": i=" << i << ", j=" << j << std::endl;
    }
}

auto t2 = high_resolution_clock::now(); // stop time measurement

In [6]:
auto duration = duration_cast<std::chrono::microseconds>(t2 - t1).count();
std::cout << "Execution time: " << duration << " microseconds" << std::endl;

Execution time: 162314 microseconds


Taking advantage of the row major memory layout makes our program **6.78x times faster** in this case!
$$162314 / 23947 \approx 6.78$$